pkg/gadgets/trace/byob/tracer/tracer.go

```diff
@@ -0,0 +1,516 @@
+				return fmt.Errorf("failed to attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		} else if p.Type == ebpf.SocketFilter && strings.HasPrefix(p.SectionName, "socket") {
+			t.socketFilterPrograms = append(t.socketFilterPrograms, progName)
+		}
+	}
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	d := t.decoderFactory()
+	ctx := context.TODO()
+	for {
+		var rawSample []byte
+
+		if t.ringbufReader != nil {
+			record, err := t.ringbufReader.Read()
+			if err != nil {
+				if errors.Is(err, ringbuf.ErrClosed) {
+					// nothing to do, we're done
+					return
+				}
+
+				msg := fmt.Sprintf("Error reading ring buffer: %s", err)
+				t.eventCallback(types.Base(eventtypes.Err(msg)))
+				return
+			}
+			rawSample = record.RawSample
+		} else if t.perfReader != nil {
+			record, err := t.perfReader.Read()
+			if err != nil {
+				if errors.Is(err, perf.ErrClosed) {
+					return
+				}
+
+				msg := fmt.Sprintf("Error reading perf ring buffer: %s", err)
+				t.eventCallback(types.Base(eventtypes.Err(msg)))
+				return
+			}
+
+			if record.LostSamples != 0 {
+				msg := fmt.Sprintf("lost %d samples", record.LostSamples)
+				t.eventCallback(types.Base(eventtypes.Warn(msg)))
+				continue
+			}
+			rawSample = record.RawSample
+		} else {
+			msg := fmt.Sprintf("Error using reader for ring buffer")
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		if uint32(len(rawSample)) < t.mapSizes[t.printMap] {
+			msg := fmt.Sprintf("Error reading ring buffer: len(RawSample)=%d!=%d",
+				len(rawSample),
+				t.mapSizes[t.printMap])
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		// FIXME: DecodeBtfBinary has a bug with non-NULL-terminated strings.
+		// For now, ensure the problem does not happen in ebpf
+
+		result, err := d.DecodeBtfBinary(ctx, t.valueStruct, rawSample[:t.mapSizes[t.printMap]])
+		if err != nil {
+			msg := fmt.Sprintf("Error decoding btf: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+		b, err := json.Marshal(result)
+		if err != nil {
+			msg := fmt.Sprintf("Error encoding json: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		event := types.Event{
+			Event: eventtypes.Event{
+				Type: eventtypes.NORMAL,
+			},
+			WithMountNsID: eventtypes.WithMountNsID{MountNsID: 0},
+			Payload:       fmt.Sprintf("%+v", string(b)),
+		}
+		var mountNsIDStruct struct {
+			MountNsID uint64 `json:"mount_ns_id"`
+		}
+		_ = json.Unmarshal(b, &mountNsIDStruct)
+		event.MountNsID = mountNsIDStruct.MountNsID
+
+		if t.enricher != nil {
+			t.enricher.EnrichByMntNs(&event.CommonData, event.MountNsID)
+		}
+
+		t.eventCallback(&event)
+	}
+}
+
+// Attach attaches networking programs in the netns related to the pid
+//
+// Tracing programs (like kprobes) are attached once during start() but
+// networking programs are attached for each netns.
+func (t *Tracer) Attach(pid uint32) (err error) {
```

Comment:
Networking gadgets like sni and dns had a Attach() method but then it was  moved to pkg/gadgets/internal/networktracer/tracer.go. I wonder if byob could reuse that instead of reimplementing the networking logic.

However, internal/networktracer is being refactored in #1430 (see also https://github.com/inspektor-gadget/inspektor-gadget/pull/1430#issuecomment-1470794513), so I think it's too much refactoring at the same time.

Maybe the first PR for the byob gadget could limit itself to tracing programs (kprobes, tracepoints). Networking support for byob could be added later, after #1430 is done. Wdyt?

Reply:
Right, then i'll keep a version of it stashed in another branch and wait for that PR